File: examples/matlab/lp/wegner2025-section-5-2.m
clear; 
clc; 

addpath("~/Documents/repos/tca-matlab-toolbox/");
addpath("~/Documents/repos/tca-matlab-toolbox/plotting");
addpath("~/Documents/repos/tca-matlab-toolbox/models");

data = readtable("./data.csv");

% 1. Defining the model 
model = LP(data, 'newsy', 4, 0:20, 'includeConstant', true);
model.fit();

% 2. Obtaining total effects
method = Recursive()
irfObj = model.IRF(20, 'identificationMethod', method);
irfs = irfObj.getIrfArray();

model.getVariableNames()
plot(0:20, vec(irfs(2, 1, :)))  % government defense spending
plot(0:20, vec(irfs(3, 1, :)))  % total government spending
plot(0:20, vec(irfs(4, 1, :)))  % output

% 3. Defining the transmission  matrix
transmissionOrder = {'newsy', 'gdef', 'g', 'y'};
transmissionOrder = model.getVariableNames();  % because order is already correct

% 4. Defining transmission channels
anticipationChannel = model.notThrough('gdef', 0:20, transmissionOrder);
implementationChannel = ~anticipationChannel;

% 5. Computing transmission effects
anticipationEffects = model.transmission(1, anticipationChannel, transmissionOrder, 20, 'identificationMethod', method);
implementationEffects = model.transmission(1, implementationChannel, transmissionOrder, 20, 'identificationMethod', method);

% 6. Visualising
channelNames = ["Anticipation Channel", "Implementation Channel"];
cellChannelEffects = {anticipationEffects, implementationEffects};
% output
fig = plotDecomposition(4, irfs(:, 1, :), cellChannelEffects, channelNames)
set(fig, 'Units', 'inches', 'Position', [1 1 10 5]);
exportgraphics(fig, 'output.png', 'Resolution', 300);
% Total government spending
fig = plotDecomposition(3, irfs(:, 1, :), cellChannelEffects, channelNames)
set(fig, 'Units', 'inches', 'Position', [1 1 10 5]);
exportgraphics(fig, 'gov.png', 'Resolution', 300);
% Government defense spending
fig = plotDecomposition(2, irfs(:, 1, :), cellChannelEffects, channelNames)
set(fig, 'Units', 'inches', 'Position', [1 1 10 5]);
exportgraphics(fig, 'gdef.png', 'Resolution', 300);

File: examples/julia/lp/wegner2025-section-5-2.jl
using TransmissionChannelAnalysis
using DataFrames, CSV
using CairoMakie  # for plotting

data = DataFrame(CSV.File("./data.csv"))

# 1. Defining the model
model = LP(data, :newsy, 4, 0:20; include_constant=true)
fit!(model)

# 2. Obtaining total effects
method = Recursive()
irfs = IRF(model, method, 20)
irfs = irfs.irfs[:, 1:1, :]

# 3. Defining the transmission matrix
transmission_order = [:newsy, :gdef, :g, :y]

# 4. Defining transmission channels
anticipation_channel = not_through(model, :gdef, 0:20, transmission_order)
implementation_channel = !anticipation_channel

# 5. Computing transmission effects
anticipation_effects = transmission(model, method, 1, anticipation_channel, transmission_order, 20)
implementation_effects = transmission(model, method, 1, implementation_channel, transmission_order, 20)

# 6. Visualising the effects
teffects = [anticipation_effects, implementation_effects]
channel_names = ["Anticipation", "Implementation"]

fig = Figure(; size=(1000, 300));
ax1 = Axis(fig[1, 1]; title="GDP")
plot_decomposition!(ax1, 4, irfs, teffects)
ax2 = Axis(fig[1, 2]; title="Total Government Spending")
plot_decomposition!(ax2, 3, irfs, teffects)
ax3 = Axis(fig[1, 3]; title="Defense Spending")
plot_decomposition!(ax3, 2, irfs, teffects)
add_decomposition_legend!(fig[2, :], channel_names)

fig

save("./decomposition.png", fig)

